File: src/visual/interactivity/notifications.py
# Copyright (c) 2022-2026 MKM Research Labs.
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

"""
User notification system for interactive map functionality.

Provides toast-style notifications for user feedback in the map interface.
"""

import json
from enum import Enum
from typing import Any, Dict

import folium

from visual.interactivity._jsbundle import js_static
from config.theme import colour


class NotificationType(Enum):
    """Notification types with associated styling."""
    INFO = ("info", "ℹ️", colour('accent-bright'), colour('accent-soft'))
    SUCCESS = ("success", "✅", colour('green-bright'), colour('flood-none'))
    WARNING = ("warning", "⚠️", colour('amber-bright'), colour('warn-bg-warm'))
    ERROR = ("error", "❌", colour('red-bright'), colour('danger-bg-soft'))
    LOADING = ("loading", "🔄", colour('grey'), colour('sunken'))

    def __init__(self, key: str, icon: str, color: str, background: str):
        self.key = key
        self.icon = icon
        self.color = color
        self.background = background


class NotificationPosition(Enum):
    """Notification container positions."""
    TOP_RIGHT = "top-right"
    TOP_LEFT = "top-left"
    TOP_CENTER = "top-center"
    BOTTOM_RIGHT = "bottom-right"
    BOTTOM_LEFT = "bottom-left"
    BOTTOM_CENTER = "bottom-center"


class NotificationSystem:
    """Toast notification system for map interactions."""

    def __init__(self,
                 position: NotificationPosition = NotificationPosition.TOP_RIGHT,
                 timeout: int = 5000,
                 max_visible: int = 5):
        """
        Initialize notification system.

        Args:
            position: Screen position for notifications
            timeout: Auto-dismiss timeout in milliseconds
            max_visible: Maximum simultaneous notifications
        """
        self.position = position
        self.timeout = timeout
        self.max_visible = max_visible

    def _build_templates_js(self) -> str:
        """Build JS object for notification templates."""
        templates = {
            t.key: {"icon": t.icon, "color": t.color, "background": t.background}
            for t in NotificationType
        }
        return json.dumps(templates)

    def _build_position_css(self) -> str:
        """Build CSS for position mapping."""
        return js_static('notifications_position.js')

    def get_js(self) -> str:
        """Generate the notification system CSS and JavaScript."""
        from config import config
        static_dir = config.get_static_dir()
        css_code = (static_dir / 'css' / 'notifications.css').read_text()
        js_code = (static_dir / 'js' / 'notifications.js').read_text()
        notif_config = json.dumps({
            'position': self.position.value,
            'timeout': self.timeout,
            'maxVisible': self.max_visible,
            'templates': {
                t.key: {"icon": t.icon, "color": t.color, "background": t.background}
                for t in NotificationType
            }
        })
        return f"<style>{css_code}</style>\n<script>window.__NOTIF_CONFIG = {notif_config};\n{js_code}</script>"

    def add_to_map(self, folium_map: folium.Map) -> None:
        """Add notification system to a Folium map."""
        folium_map.get_root().html.add_child(folium.Element(self.get_js()))

    def configure(self, position: NotificationPosition = None,
                 timeout: int = None, max_visible: int = None) -> None:
        """Update configuration."""
        if position:
            self.position = position
        if timeout is not None:
            self.timeout = timeout
        if max_visible is not None:
            self.max_visible = max_visible

    def get_statistics(self) -> Dict[str, Any]:
        """Get configuration statistics."""
        return {
            'default_position': self.position.value,
            'auto_dismiss_timeout': self.timeout,
            'max_notifications': self.max_visible,
            'available_types': [t.key for t in NotificationType],
            'available_positions': [p.value for p in NotificationPosition]
        }


# Convenience factory functions
def create_notification_system(position: str = "top-right",
                               timeout: int = 5000,
                               max_notifications: int = 5) -> NotificationSystem:
    """Create a notification system with common settings."""
    try:
        pos = NotificationPosition(position)
    except ValueError:
        pos = NotificationPosition.TOP_RIGHT

    return NotificationSystem(position=pos, timeout=timeout, max_visible=max_notifications)


def add_notifications_to_map(folium_map: folium.Map, **kwargs) -> NotificationSystem:
    """Add notifications to a map (convenience function)."""
    system = create_notification_system(**kwargs)
    system.add_to_map(folium_map)
    return system

File: src/static/js/notifications.js
// Copyright (c) 2022-2026 MKM Research Labs.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

(function(root) {
    'use strict';

    function createNotificationSystem(config) {
        config = config || {};
        var position = config.position || 'top-right';
        var timeout = config.timeout != null ? config.timeout : 5000;
        var maxVisible = config.maxVisible || 5;
        var templates = config.templates || {
            info:    { icon: '\u2139\uFE0F', color: Theme.value('accent-bright'), background: Theme.value('accent-soft') },
            success: { icon: '\u2705',       color: Theme.value('green-bright'), background: Theme.value('ok-bg') },
            warning: { icon: '\u26A0\uFE0F', color: Theme.value('amber-bright'), background: Theme.value('warn-bg-warm') },
            error:   { icon: '\u274C',       color: Theme.value('red-bright'), background: Theme.value('danger-bg-soft') },
            loading: { icon: '\uD83D\uDD04', color: Theme.value('grey'), background: Theme.value('sunken') }
        };

        var queue = [];
        var counter = 0;

        function getContainer() {
            var doc = root.document;
            var c = doc.getElementById('notif-container');
            if (!c) {
                c = doc.createElement('div');
                c.id = 'notif-container';
                c.className = position || 'top-right';
                doc.body.appendChild(c);
            }
            return c;
        }

        function show(message, type, options) {
            type = (type || 'info').toLowerCase();
            options = options || {};
            var tpl = templates[type] || templates.info;
            var id = ++counter;

            queue.push(id);
            while (queue.length > maxVisible) {
                dismiss(queue.shift());
            }

            var doc = root.document;
            var el = doc.createElement('div');
            el.id = 'notif-' + id;
            el.className = 'notif';
            el.style.background = tpl.background;
            el.style.borderLeftColor = tpl.color;

            el.innerHTML =
                '<div class="notif-content">' +
                '<span class="notif-icon">' + tpl.icon + '</span>' +
                '<div class="notif-message">' + message.replace(/\n/g, '<br>') + '</div>' +
                '<button class="notif-close" onclick="window.dismissNotification(' + id + ')">&times;</button>' +
                '</div>';

            el.addEventListener('click', function(e) {
                if (e.target.tagName !== 'BUTTON') dismiss(id);
            });

            getContainer().appendChild(el);
            setTimeout(function() { el.classList.add('show'); }, 10);

            if (!options.persistent && timeout > 0) {
                setTimeout(function() { dismiss(id); }, options.timeout || timeout);
            }

            return id;
        }

        function dismiss(id) {
            var doc = root.document;
            var el = doc.getElementById('notif-' + id);
            if (el) {
                el.classList.remove('show');
                setTimeout(function() { if (el.parentNode) el.parentNode.removeChild(el); }, 300);
            }
            queue = queue.filter(function(n) { return n !== id; });
        }

        function update(id, message, type) {
            var doc = root.document;
            var el = doc.getElementById('notif-' + id);
            if (!el) return;
            var textEl = el.querySelector('div > div');
            if (textEl) textEl.innerHTML = message.replace(/\n/g, '<br>');
            if (type) {
                var tpl = templates[type] || templates.info;
                var iconEl = el.querySelector('span');
                if (iconEl) iconEl.textContent = tpl.icon;
                el.style.background = tpl.background;
                el.style.borderLeftColor = tpl.color;
            }
        }

        function clearAll() {
            queue.forEach(dismiss);
            queue = [];
        }

        function getQueue() { return queue.slice(); }

        return {
            show: show,
            dismiss: dismiss,
            update: update,
            clearAll: clearAll,
            getQueue: getQueue
        };
    }

    // Node.js / Jest export
    if (typeof module !== 'undefined' && module.exports) {
        module.exports = { createNotificationSystem: createNotificationSystem };
    }

    // Browser auto-init
    if (typeof root !== 'undefined' && root.document && root.__NOTIF_CONFIG) {
        var api = createNotificationSystem(root.__NOTIF_CONFIG);
        root.showNotification = api.show;
        root.dismissNotification = api.dismiss;
        root.updateNotification = api.update;
        root.clearAllNotifications = api.clearAll;
        root.showSuccess = function(msg, opts) { return api.show(msg, 'success', opts); };
        root.showError = function(msg, opts) { return api.show(msg, 'error', Object.assign({persistent: true}, opts || {})); };
        root.showWarning = function(msg, opts) { return api.show(msg, 'warning', opts); };
        root.showInfo = function(msg, opts) { return api.show(msg, 'info', opts); };
        root.showLoading = function(msg, opts) { return api.show(msg, 'loading', Object.assign({persistent: true}, opts || {})); };
        if (root.console) root.console.log('Notification system initialized');
    }

})(typeof window !== 'undefined' ? window : (typeof global !== 'undefined' ? global : this));
